### ISSUE TITLE
wrong symbol address resolving for arm THUMB addressing mode

### ISSUE THREAD

**Issue by bieganski** [libbpf/libbpf#858]
as the title says, `libbpf`'s symbol address resolver does not take into account THUMB addressing mode.

```bash
m.bieganski@hostname:~$ file libssl.so.1.1 
libssl.so.1.1: ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, stripped
m.bieganski@hostname:~$ 
m.bieganski@hostname:~$ 
m.bieganski@hostname:~$ nm -D libssl.so.1.1 | grep SSL_read@
000221e5 T SSL_read@@OPENSSL_1_1_0
m.bieganski@hostname:~$ 
m.bieganski@hostname:~$ 
m.bieganski@hostname:~$ arm-linux-gnueabi-nm -D libssl.so.1.1 | grep SSL_read@
000221e4 T SSL_read@@OPENSSL_1_1_0
```

The `libbpf`'s resolver's behavior resembles my local `nm`, also being not aware of THUMB addressing convention, to use jump target's LSB to specify THMB mode on/off. Apparently this information bit is also present in ELF's symbol table, as parsed by `nm`.  For libbpf implementation see https://github.com/libbpf/libbpf/blob/master/src/elf.c#L269.

As a consequence the kernel is unable to set the (u)kprobe and `perf_event_open` fails (see check at https://github.com/torvalds/linux/blob/master/kernel/events/uprobes.c#L1204)

similar issue:
https://stackoverflow.com/questions/67156676/elf-file-read-symbol-and-symbol-address-1-byte-difference

**Comment by anakryiko** [comment]
@bieganski do you have a link to some spec explaining what this is and how it should be handled? Seems like the fix is to clear lowest bit if ELF binary is THUMB, right? But it would be nice to have some spec describing all the magical piece in one place, if you have that. Thanks!

**Comment by tnovak** [comment]
> perf_event_open fails at a further stage, since THUMB is not supported in current kernel

Right, the kernel doesn't (yet) support probing Thumb code. At the very least, I believe `set_swbp()` would need to be extended to support both 16- and 32-bit Thumb encodings of breakpoint instructions (UDF), since it's different from the ARM variant.

From libbpf's perspective, I think it makes sense to preserve the current behavior (least significant set if the address points to a Thumb instruction) -- it's consistent with [AAPCS32](https://github.com/ARM-software/abi-aa/blob/main/aapcs32/aapcs32.rst#51fundamental-data-types) and will allow the kernel to distinguish between ARM and Thumb instruction sets.
